community/kernel/src/main/java/org/neo4j/kernel/api/ExecutingQuery.java

```diff
@@ -20,41 +20,64 @@
 package org.neo4j.kernel.api;
 
 import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
 
+import org.apache.commons.lang3.builder.ToStringBuilder;
+
+import org.neo4j.kernel.impl.locking.Locks;
 import org.neo4j.kernel.impl.query.QuerySource;
+import org.neo4j.storageengine.api.lock.ResourceType;
+import org.neo4j.time.CpuClock;
+import org.neo4j.time.SystemNanoClock;
 
-import static java.lang.String.format;
+import static java.util.concurrent.atomic.AtomicLongFieldUpdater.newUpdater;
 
 /**
  * Represents a currently running query.
  */
 public class ExecutingQuery
 {
+    private static final AtomicLongFieldUpdater<ExecutingQuery> WAIT_TIME =
+            newUpdater( ExecutingQuery.class, "waitTimeNanos" );
     private final long queryId;
+    private final Locks.Tracer lockTracer = ExecutingQuery.this::waitForLock;
```

Comment:
Tracer implementations are meant to be changed via plug-ins, and the concrete implementation decided by `Tracers`. Hard coding the implementation here goes against this. For this reason, tracers should also not be relied upon for correctness; which is another mistake we've made with monitors. I think you should allow `Tracers` to decide what implementation to use, and then you can move the clock state into the default implementation. I would not be against adding a method to the interface to query the accumulated wait time, as long as you can accept that an alternative implementation might not track it and instead always report zero.